code-demos/code_demo_logreg.Rmd

```diff
@@ -0,0 +1,431 @@
+- Note: The classic Newton-Raphson method uses $\lambda=1$, but we can enlarge its radius of convergence by optimizing $\lambda$ in every step.
+
+We can implement this quite easily by realizing that we are principally doing the same as with the gradient descent method - except that
+now we use a "weighted" gradient.
+
+```{r logreg-newton}
+library(MASS)
+
+newton_opt_stepsize <- function(Y, X, theta,
+                                       risk,
+                                       gradient,
+                                       hessian,
+                                       lambda = 0.005,
+                                       epsilon = 0.0001,
+                                       max_iterations = 2000,
+                                       min_learning_rate = 0,
+                                       max_learning_rate = 1000,
+                                       plot = TRUE,
+                                       include_storage = FALSE) {
+gradient_descent_opt_stepsize(Y, X, theta,
+                                       risk = risk,
+                                       gradient = function(Y,X,theta)
+                                         ginv(hessian(Y,X,theta)) %*%
+                                         gradient(Y,X,theta),
+                                       lambda = lambda,
+                                       epsilon = epsilon,
+                                       max_iterations = max_iterations,
+                                       min_learning_rate = min_learning_rate,
+                                       max_learning_rate = max_learning_rate,
+                                       plot = plot,
+                                       include_storage = include_storage
+                                      )
+}
+```
+
+
+## Idea of Newton-Raphson
+
+```{r logreg_idea_newton, echo=FALSE}
+num_data <- 100
+
+x <- seq(from = 0,to =7,length.out = 100)
+
+f <- function(x) x^2 - 1
+fx <- function(x) 2*x
+
+func_data <- data.frame(x = x, y = f(x))
+
+x0 <- 6.5
+slope <- fx(x0)
+intercept <- f(x0) - slope*x0
+
+x1 <- x0 - slope^-1*f(x0)
+
+ggplot() + 
+  geom_line(data=func_data,aes(x=x,y=y), color="purple") +
+  geom_hline(yintercept = 0) + 
+  geom_vline(xintercept = 6.5, color="yellow") + 
+  geom_abline(slope=slope, intercept = intercept, color="turquoise") + 
+  geom_text(aes(label='x^{~"[1]" }',x=x1, y=-1.5),parse = TRUE) +
+  geom_text(aes(label='x^{~"[0]" }',x=x0, y=-1.5),parse = TRUE) +
+  geom_text(aes(label='"f\'"(x^{~"[0]" })',x=x0+0.3, y=f(x0)),parse = TRUE,
+            color="turquoise") + 
+  geom_text(aes(label='f(x)',x=4, y=20),parse = TRUE,
+            color="purple")
+```
+
+If our scalar function $f$ of interest has a root and we are *near enough* to it, we can 
+find it by iteratively using following procedure:
+$$
+x^{[t]} = x^{[t-1]} - f'(x^{[t-1]})\cdot f(x^{[t-1]}).
```

Comment:
- needs inverse first derivative. 
- add some explanations to formula? or find good explainer (video) on newton's method online and link it
- needs additional explanations to make connections clearer:
  - minimizing emprical risk = root finding of derivative of emprical risk
  - so f <-> derivative of empirical risk, f' <-> hessian of R_emp(theta)

Reply:
you aluude to this below, but should be made more explicit. i think